notebook/exploration-reports/2022.03-ipld-url-scheme.md

```diff
@@ -0,0 +1,111 @@
+IPLD URL Scheme v0
+==================
+
+This is a proposed IPLD URL scheme spec, focusing on a minimal viable specification which doesn't address advanced use cases such as lenses or selectors.
+
+**Author**: Mauve Signeaver ([@RangerMauve])
+
+---
+
+tl;dr
+
+```
+GET ipld://bafy....etc/some/path => {some: data}
+PUT ipld://bafy....etc/some/path {some: data} => ipld://bafy...etc/
+```
+
+## Objective
+
+`ipfs://` exists out in the wild to provide ways to link to files, and some web browsers integrate IPFS URLs to load and author content.
+This spec shows how a similar URL scheme can exist for IPLD data.
+Part of the goal of the spec is to make it easy to reuse more or less the same APIs between software that handles `ipld://` URLs natively, HTTP gateways like the [IPFS gateway](https://docs.ipfs.io/concepts/ipfs-gateway/) and CLIs.
+With that in mind, there should be a clear mapping between HTTP-like APIs via gateways and URL handlers, and CLI arguments.
+In order to make it easier to come to an initial spec, we will focus on a minimal set of features and make subsequent specs with support for features like ADLs, Schemas, and advanced selectors.
```

Comment:
While I get the temptation to just do this a little bit at a time, I think we should do it with our eyes towards the future.

In particular, my concern is that `ipld://` as described is almost exactly the same as `ipfs://` and has limited utility in that there doesn't seem to be a way to include ADLs or selectors in a way that wouldn't also work for `ipfs://`. This means that what we get out of `ipld://` in comparison to `ipfs://` is a fairly limited domain (and is what @rvagg alluded to with how `/ipld` currently works in some go-ipfs commands) that AFAICT mostly boils down to debugging UnixFS (at least on the GET side). So why bother asking implementers (e.g. browsers) to support both `ipfs://` and `ipld://`?

In a recent experiment with using selectors with `ipfs://` I went with `?selector=` which also allowed me to utilize ADLs (code starts here https://github.com/aschmahmann/go-ipfs/blob/b0859a906110bdac348d2888dfd6ddd960df738e/core/corehttp/gateway_handler.go#L388). `ipfs://<cid>?selector=<selector>?format=<some format>` seems like it would cover almost all of the GET use case described here for `ipld://`.

I wonder if we could help ourselves out here with some type of escaping in `ipld://` that will allow us to do more interesting things that we can't in `ipfs://` for legacy reasons (i.e. no reserved characters). For example: `ipld://bafyfoobar/[ADL=MyHAMTv1]/field1/[ADL=YourHAMTv2]/field2/[ADL=FBL]/[Selector=Slice[100,400]]` might be more readable than some base-encoded inline selector or CID of a selector for the cost of disallowing `[...]` as path segments (or require escaping in that case)

Reply:
That's a really good point regarding the existing IPFS gateway APIs already being able to support most of the `GET` use cases. I was imagining that that would change more as time went one. I think the authoring of IPLD data with `PUT` was more interesting to me as a differentiator since the writable gateway APIs are currently very file-focused and it'd feel super hacky to add a new mode to them for IPLD with `?format`.

Authoring IPLD data as well as consuming it with IPLD-specific semantics is the main motivator for having this be a separate URL scheme / gateway path for me.

I really like the idea of having extra parameters within individual segments of the path!
I think this could also help with the stuff @rvagg mentioned about type hints for individual nodes that get auto-created when you're using `PUT` to add a node to a given path.

From your example URL, the format seems to be `ipld://CID/SEGMENT_OR_SELECTORS`, but what if it we added the selectors onto the segment itself?

e.g. `PUT ipld://bafyduck/field1[Kind=Map]/field2`?

Reply:
_If_ you go with special characters, keep in mind possible escaping options--we face critique in a few of our specs for the lack of escaping to allow the full domain of fields/names/etc. Pathing is particularly problematic in this respect since it is already restricted in the domain of field names. Taking `[` & `]` off the board as well would be unfortunate (although not the worst thing, but we do tend to see a bit of `[`, `]` use for field names in data mainly thanks to web tooling conventions). I don't really want to see this descend into a rabbit hole of optionality but a bit of forward thinking might be helpful.